kivy/uix/textinput.py

```diff
@@ -2571,6 +2770,7 @@ def keyboard_on_key_down(self, window, keycode, text, modifiers):
             self.focus = False
             return True
         elif key == 9:  # tab
+            self.delete_selection()
```

Comment:
What puzzled me is that we insert tab, even if readonly. However, I guess `insert_text` returns early if readonly, but still it would be nice if it was guarded with a check for readonly. But it's not very important.

The other thing that puzzled me is that editable includes when it is disabled. But users may not want disabled text to be interactive, including copying. But, I guess when it's disabled all touch is disabled for the widget anyway.

Also, it was confusing to me that we explicitly allow ctrl-a when readonly from `_handle_shortcut`. But, what about up-arrow scrolling. But I guess that is handled in `_key_down` seperately.

Reply:
yes, there are various handling for various kind of typi typi, and it's not clear to me why we segregate "shortcuts", "interesting keys", and input, it makes *some* sense, but it's still a bit arbitrary.
To avoid testing in multiple places when functions are called, i did move the test inside the function, the performance cost seems negligible as users don't type thousands of keys per second, and the benefit is that you don't have to wonder if you fill all the conditions to call it, it's its responsability to early exit if not, but i guess it could be considered a smell, considering the complexity of this widget, i'd rather move downward when possible.

Reply:
just to check i made a TextInput where disabled is based on time, so it "blinks" disabled, and even if it had focus before being disabled, it's not possible to type into it when it gets disabled, it also loses focus in that situation, which i'm not sure should be considered a bug or not.